bcm27xx/bcm2712: add RP1 functionality

graysky2 opened the pull request:
Add RP1 features as built-ins and give users the option to build some more niche ones as kmods:
* Composite video
* Display video
* LED control
* PWM control
* Serial video

Build system: x86/64
Build-tested: bcm2712/RPi5B
Run-tested: bcm2712/RPi5B

graysky2 committed:
bcm27xx/bcm2712: add RP1 drivers via kmods

Add kmods for the following RP1 options that not all users
will necessarily need or want compiled in:
* Composite video
* Display video
* LED control
* PWM control
* Serial video

Build system: x86/64
Build-tested: bcm2712/RPi5B
Run-tested: bcm2712/RPi5B

Signed-off-by: John Audia <therealgraysky@proton.me>
Link: https://github.com/openwrt/openwrt/pull/17233
Signed-off-by: Robert Marko <robimarko@gmail.com>

graysky2 commented:
Seeking advise on how integrate these into I am assuming the various files within `package/kernel/linux/modules/*.mk`

graysky2 commented:
Just realized that I had this PR applied to my https://github.com/graysky2/openwrt/tree/6.12/bcm27xx/bcm2712 branch and it builds and functions as expected there, but when I compile against HEAD, none of the new modules are generated and their corresponding packages are empty.  What am I missing?

```
% find build_dir/target-aarch64_cortex-a76_musl -name '*rp1*.ko'
```

```
% ls -lh bin/targets/bcm27xx/bcm2712/packages/kmod-rp1*
-rw-r--r-- 1 facade users 723 Dec 12 11:47 bin/targets/bcm27xx/bcm2712/packages/kmod-rp1-firmware-6.6.63-r1.apk
-rw-r--r-- 1 facade users 719 Dec 12 11:47 bin/targets/bcm27xx/bcm2712/packages/kmod-rp1-mailbox-6.6.63-r1.apk
-rw-r--r-- 1 facade users 720 Dec 12 11:47 bin/targets/bcm27xx/bcm2712/packages/kmod-rp1-pio-6.6.63-r1.apk
```

leobsky commented:
Missing the latest RPi patches, haven't been ported to OpenWrt https://github.com/raspberrypi/linux/commit/0d58d8cfb6f989f290d983552fcaa116e582e84a

graysky2 commented:
Thanks @leobsky - I added that patch which built the first module... how did you locate it?  I would like to identify the needed cherry picks for the remaining 3.

graysky2 committed:
bcm27xx: patches: cherry-pick for RP1 kmods

Cherry-pick patches to support building RP1 modules.

Signed-off-by: John Audia <therealgraysky@proton.me>
Link: https://github.com/openwrt/openwrt/pull/17233
Signed-off-by: Robert Marko <robimarko@gmail.com>

leobsky commented:
> EDIT: something must not be right.... when I run the following in an attempt to just refresh the bcm27xx patches against rpi foundation's fork of the kernel, I get over 13,000 of them...
> 
> ```
> git format-patch v6.6.3..rpi/rpi-6.6.y
> ```

@graysky2 You can also use the the hash of the latest merged kernel in the the RPi kernel to generate the patches

`git format-patch --no-merges -N rpi-6.6...22a054ea1f081d7837cc8e24ad4c7aa36e8bba04`

graysky2 commented:
Ah, I had a type on ... comparing version 6.6.3 not 6.6.**6**3

graysky2 commented:
Some checks are failing as expected for targets that are not bcm2712 ... what is the right way to scope the build to only the bc27xx/bcm2712?

namiltd commented:
I think that when there is @TARGET_bcm27xx_bcm2712 then @!TARGET_bcm27xx-bcm(2708||2709||2710||2711) is unnecessary.

graysky2 commented:
@namiltd - I tried it both way but still the CI is building for the other subtargets.

namiltd commented:
A dash instead of an underscore in @!TARGET_bcm27xx**-**bcm(2708||2709||2710||2711)

namiltd commented:
Try no
@!TARGET-bcm27xx-bcm(2708||2709||2710||2711)
but
@!TARGET-bcm27xx_bcm(2708||2709||2710||2711)

hauke commented on target/linux/bcm27xx/modules/other.mk:
```diff
@@ -53,3 +53,63 @@ define KernelPackage/smi-bcm2835-dev/description
+           CONFIG_FIRMWARE_RP1 \
+           CONFIG_MBOX_RP1
+  FILES:=$(LINUX_DIR)/drivers/firmware/rp1.ko
+  DEPENDS:=@!TARGET_bcm27xx_bcm(2708||2709||2710||2711) @TARGET_bcm27xx
```
Please write the full depends like like:
```suggestion
  DEPENDS:=@!TARGET_bcm27xx_bcm2708 @!TARGET_bcm27xx_bcm2709 @!TARGET_bcm27xx_bcm2710 @!TARGET_bcm27xx_bcm2711  @TARGET_bcm27xx
```
Maybe just making it depend on `@!TARGET_bcm27xx_bcm2712` is better.

graysky2 replied:
@hauke - I trying your suggestion of the full depends now.  As to your 2nd suggestion, `@!TARGET_bcm27xx_bcm2712`, would make it build for 8,9,10, and 11 but we want it to only build for 12.

graysky2 commented:
@leobsky - I'd like to also add the drm-rp1 modules, `drm-rp1-dsi` `drm-dpi-dpi` and `drm-rp1-vec`.  I started with just one but am finding the same "module is missing" error (see below)  To help control for this, I first applied https://github.com/openwrt/openwrt/pull/17147 which includes nearly all of the RPi Foundations patches for the 6.12.y branch but am still getting that error.  Wondering if you have any ideas?

```
diff --git a/target/linux/bcm27xx/modules/video.mk b/target/linux/bcm27xx/modules/video.mk
index b9af0b3e0f86..955cb198f8b0 100644
--- a/target/linux/bcm27xx/modules/video.mk
+++ b/target/linux/bcm27xx/modules/video.mk
@@ -115,3 +115,24 @@ define KernelPackage/vchiq-mmal-bcm2835/description
 endef
 
 $(eval $(call KernelPackage,vchiq-mmal-bcm2835))
+
+
+define KernelPackage/drm-rp1-dsi
+  SUBMENU:=$(VIDEO_MENU)
+  TITLE:=RP1 Display Serial Interface for Video
+  KCONFIG:=CONFIG_DRM_RP1_DSI
+  FILES:=$(LINUX_DIR)/drivers/gpu/drm/dp1/rp1-dsi/drm-rp1-dsi.ko
+  AUTOLOAD:=$(call AutoLoad,67,drm-rp1-dsi)
+  DEPENDS:= \
+   @!TARGET_bcm27xx_bcm2708 @!TARGET_bcm27xx_bcm2709 \
+   @!TARGET_bcm27xx_bcm2710 @!TARGET_bcm27xx_bcm2711 \
+   @TARGET_bcm27xx +kmod-drm-vc4
+endef
+
+define KernelPackage/drm-rp1-dsi/description
+  This module manages the DSI for driving high-resolution LCD panels
+  such as the official Raspberry Pi displays or other screens that
+  use the DSI interface.
+endef
+
+$(eval $(call KernelPackage,drm-rp1-dsi))
```

Here is the error:
```
...
/scratch/union/staging_dir/host/bin/fakeroot /scratch/union/staging_dir/host/bin/apk mkpkg --info "name:kmod-drm-rp1-dsi" --info "version:6.12.5-r1" --info "description:This module manages the DSI for driving high-resolution LCD panels such as the official Raspberry Pi displays or other screens that use the DSI interface." --info "arch:aarch64_cortex-a76" --info "license:GPL-2.0-only" --info "origin:feeds/base/kernel/linux" --info "url:" --info "maintainer:" --info "provides:"  --script "post-install:/scratch/union/build_dir/target-aarch64_cortex-a76_musl/linux-bcm27xx_bcm2712/packages/apk-aarch64_cortex-a76/kmod-drm-rp1-dsi/post-install" --script "pre-deinstall:/scratch/union/build_dir/target-aarch64_cortex-a76_musl/linux-bcm27xx_bcm2712/packages/apk-aarch64_cortex-a76/kmod-drm-rp1-dsi/pre-deinstall" --info "depends:kernel=6.12.5~7d9783d00a291abf18c479b0c0ef4d3c-r1 kmod-rp1-firmware kmod-drm-vc4" --files "/scratch/union/build_dir/target-aarch64_cortex-a76_musl/linux-bcm27xx_bcm2712/packages/ipkg-aarch64_cortex-a76/kmod-drm-rp1-dsi" --output "/scratch/union/bin/targets/bcm27xx/bcm2712/packages/kmod-drm-rp1-dsi-6.12.5-r1.apk" --sign "/scratch/union/private-key.pem"
[ -f /scratch/union/bin/targets/bcm27xx/bcm2712/packages/kmod-drm-rp1-dsi-6.12.5-r1.apk ]
ERROR: module '/scratch/union/build_dir/target-aarch64_cortex-a76_musl/linux-bcm27xx_bcm2712/linux-6.12.5/drivers/gpu/drm/dp1/rp1-dsi/drm-rp1-dsi.ko' is missing.
make[3]: *** [/scratch/union/target/linux/bcm27xx/modules/video.mk:138: /scratch/union/bin/targets/bcm27xx/bcm2712/packages/kmod-drm-rp1-dsi-6.12.5-r1.apk] Error 1
make[3]: Leaving directory '/scratch/union/package/kernel/linux'
time: package/kernel/linux/compile#5.89#1.48#6.60
    ERROR: package/kernel/linux failed to build.
make[2]: *** [package/Makefile:185: package/kernel/linux/compile] Error 1
```

graysky2 committed:
kernel: bcm27xx/bcm2712: add RP1 functionality

Build in several options RP1-specific features rather than
generating additional kmods for them since bcm2712 is unique to
RPi5B only.

Signed-off-by: John Audia <therealgraysky@proton.me>
Link: https://github.com/openwrt/openwrt/pull/17233
Signed-off-by: Robert Marko <robimarko@gmail.com>

graysky2 commented:
@Noltari - I did not realize there are no other bcm2712 SoCs.  I think building several of these into the kernel is a good idea but  don't think right for all of them since some users will not need all of them.  For example the PWM, and the video ones I am trying to add not.  What do you think?

This PR just adds the missing patches relating to RP1 since openwrt/main contains the rest of the RPi patches.

graysky2 commented:
I am getting the following error for the camera front-end:
```
    ERROR: module '/scratch/union/build_dir/target-aarch64_cortex-a76_musl/linux-bcm27xx_bcm2712/linux-6.6.66/drivers/media/common/videobuf2/videobuf2-vmalloc.ko' is missing.
    make[3]: *** [modules/video.mk:681: /scratch/union/bin/targets/bcm27xx/bcm2712/packages/kmod-video-videobuf2-6.6.66-r1.apk] Error 1
    make[3]: Leaving directory '/scratch/union/package/kernel/linux'
    time: package/kernel/linux/compile#5.95#1.54#6.74
        ERROR: package/kernel/linux failed to build.
    make[2]: *** [package/Makefile:185: package/kernel/linux/compile] Error 1
```

I can't seem to get the depends quite right.  Suggestions are appreciated.

@namiltd @Noltari @leobsky

graysky2 commented:
As a test, I added `CONFIG_VIDEO_RP1_CFE=y` to `target/linux/bcm27xx/bcm2712/config-6.6` and the build was successful so I am wondering if the issues with the module stem from 1, me using the dependencies incorrectly or 2. a bug with our packaging of those dependencies.

namiltd commented:
Add `# CONFIG_RP1_PIO is not set' to target/linux/bcm27xx/config-6.6

graysky2 commented:
Thanks @namiltd  but it seems that the CI still wants to build the other targets

namiltd commented:
CI has a strange logic though.

graysky2 commented:
@namiltd - nice, the checks are passing.  any thoughts on why the cfe piece is not building?

graysky2 commented:
I think this PR is good to merge as-is based on my testing.  Anyone object?

robimarko commented:
Thanks! Rebased on top of main and merged!
